src/Polysemy/State.hs

```diff
@@ -157,6 +162,59 @@ stateToIO s sem = do
   return (end, res)
 {-# INLINE stateToIO #-}
 
+------------------------------------------------------------------------------
+-- | Run a 'State' effect by transforming it into operations over an 'STRef'.
+--
+-- @since TODO: version
+runStateSTRef
+    :: forall s st r a
+     . Member (Embed (ST st)) r
+    => STRef st s
+    -> Sem (State s ': r) a
+    -> Sem r a
+runStateSTRef ref = interpret $ \case
+  Get   -> embed $ readSTRef ref
+  Put s -> embed $ writeSTRef ref s
+{-# INLINE runStateSTRef #-}
+
+--------------------------------------------------------------------
+-- | Run an 'State' effect in terms of operations
+-- in 'ST'.
+--
+-- Internally, this simply creates a new 'STRef', passes it to
+-- 'runStateSTRef', and then returns the result and the final value
+-- of the 'STRef'.
+--
+-- /Beware/: As this uses an 'STRef' internally,
+-- all other effects will have local
+-- state semantics in regards to 'State' effects
+-- interpreted this way.
+-- For example, 'Polysemy.Error.throw' and 'Polysemy.Error.catch' will
+-- never revert 'put's, even if 'Polysemy.Error.runError' is used
+-- after 'stateToST'.
+--
+-- Additionally, one has to introduce the @st@ type to 'stateToST' by binding
+-- the type after 'runM' with an explicit forall with @-XScopedTypeVariables@.
```

Comment:
Preface this by saying that this isn't necessary if you're using the plugin, like:
`Additionally, if you're not using `polysemy-plugin`, you have to...`

Reply:
done